ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs

```diff
@@ -150,9 +152,26 @@ getIsInvalidBlock ::
      forall m blk. (IOLike m, HasHeader blk)
   => ChainDbEnv m blk
   -> STM m (WithFingerprint (HeaderHash blk -> Maybe (InvalidBlockReason blk)))
-getIsInvalidBlock CDB{..} =
+getIsInvalidBlock CDB{..} = do
   fmap (fmap (fmap invalidBlockReason) . flip Map.lookup) <$> readTVar cdbInvalid
 
+
+getAskBlockValidity ::
+  forall m blk. (IOLike m, HasHeader blk)
+  => ChainDbEnv m blk
+  -> STM m (WithFingerprint (RealPoint blk -> Maybe (BlockValidity blk)))
+getAskBlockValidity CDB{..} = do
+  prevApplied <- LgrDB.getPrevApplied cdbLgrDB
+  invalid     <- readTVar cdbInvalid
+  let fn = \pt@(RealPoint _ hash) ->
+        case Map.lookup hash $ forgetFingerprint invalid of
+          Just invalidBlockInfo -> Just $ Invalid (invalidBlockReason invalidBlockInfo)
+          Nothing -> if Set.member pt prevApplied
+            then Just Valid
+            else Nothing
+  pure $ fmap (const fn) invalid
```

Comment:
So this `fmap :: (a -> b) -> WithFingerprint a -> WithFingerprint b` is the one thing that makes the fingerprint change when the `cdbInvalid` value changes.

Do we really wait on this? Or is the `StrictTVar m (WithFingerprint (InvalidBlocks blk))` even redundant and it could just be `StrictTVar m (InvalidBlocks blk)`?